Associate pairs with groups

ericworkman opened the pull request:
@mbramson 
Depends on and merges into #81 . Pairs now depend on groups. I had to change a lot of redirects and tests for this. I created a conflict with "create_pair/2" that was resolved by using the full namespace. Most back links now redirect to the profile page. I added a link to the table of groups on the profile page for the pairs for each group.

I'll probably work on beatifying some of these things soon sans react.

- [x] New /pairs
- [x] Authz for /group/:id/pairs...

mbramson commented on test/support/test_helper.ex:
```diff
@@ -55,9 +55,16 @@ defmodule Pairmotron.TestHelper do
+      |> Repo.insert!
+    UserGroup.changeset(%UserGroup{}, %{user_id: user.id, group_id: group.id}) |> Repo.insert!
+    group
+  end
```
Everything here can be replaced with ex_machina in the tests.

mbramson commented on test/controllers/pair_controller_test.exs:
```diff
@@ -1,110 +1,108 @@
+    test "pairs two users together", %{conn: conn, logged_in_user: user1, group: group} do
       user2 = insert(:user)
-      conn = get(conn, "/pairs")
+      UserGroup.changeset(%UserGroup{}, %{user_id: user2.id, group_id: group.id}) |> Repo.insert!
```
Instead of doing an insert like this, you can just do `user2 = insert(:user, %{group: group})` instead of `user2 = insert(:user)`

mbramson commented on test/controllers/pair_controller_test.exs:
```diff
@@ -1,110 +1,108 @@
       user2 = insert(:user)
-      pair = create_pair([user, user2])
-      delete conn, pair_path(conn, :delete, year, week)
+      UserGroup.changeset(%UserGroup{}, %{user_id: user2.id, group_id: group.id}) |> Repo.insert!
```
Same as above, this can be accomplished more elegantly with `user2 = insert(:user, %{group: group})`

mbramson commented on web/controllers/pair_controller.ex:
```diff
@@ -1,52 +1,55 @@
+  def index(conn, %{"group_id" => g}) do
     {year, week} = Timex.iso_week(Timex.today)
-    pairs = fetch_or_gen(year, week)
+    {gid, _} = g |> Integer.parse
```
Why not just spell out `group_id`?

mbramson commented on web/router.ex:
```diff
@@ -45,18 +45,18 @@ defmodule Pairmotron.Router do
   scope "/", Pairmotron do
     pipe_through [:browser, :authenticate] # Use the default browser stack
 
-    get "/pairs", PairController, :index
```
So see https://github.com/ElixirCLE/pairmotron/issues/83, but I don't think that we should actually get rid of the `/pairs` route, I just think that it should be repurposed. I think that all of these redirects that are now redirecting to the profile are the wrong course of action. I still think that the user should see pairs in these cases, not their profile as pairs is really the central function of the app.

So I'd make the `/pairs` index just redirect to `/groups/:group_id/pairs` where the group_id at this point is just the group that the user is a member of with the lowest id (for now). In the future it'd be nice to let the user set which group they'd like to default to.

This also lowers the amount of changes you have to make in this PR (although I realize you've already made them).

ericworkman replied:
I don't like having `/pairs` redirect to one of the user's group pair pages. I think its surprising in the best case. It would be less surprising and more useful to have `/pairs` include the pairs for each of the user's groups for that time period. I'm not going to implement that unless we both think its a good idea.

mbramson reviewed (changes requested):
Some thoughts on this PR. What do you think about the /pairs route?

ericworkman committed:
Associate pairs with group

ericworkman commented:
@mbramson Made a lot of the changes you requested. Still open on the usage of `/pairs`.

ericworkman commented:
Discussed this with @mbramson. As part of this PR, I will be making a new "/pairs" that will display the current user's pair within each of their groups. Then most of the redirect changes will change back to redirects to "/pairs". I will also be adding an authorization check to the group pairs route to verify that a user is in the group they are trying to view.

ericworkman committed:
Split PairController into GroupPairController

Split GroupPairController off of PairController.
Move pair rendering things into SharedView.
Move pair fetching and generating into PairMaker lib.
Correct controller tests.

ericworkman commented:
Still doing:

- [x] Authorization check for GroupPairController
- [x] Switch most of the redirects back to /pairs
- [x] Refactor some of the shared things between GroupPairController and PairController
- [x] Refactor PairController to remove duplicate logic between show and index
- [x] Move PairMaker specific tests out of the controller tests
- [x] Double check the need for any migrations for existing pairs
- [x] Fix issue with /pairs showing more than its supposed to

ericworkman committed:
Change some redirects to profile back to pairs

ericworkman committed:
Add authorization to group pairs controller

ericworkman committed:
Deduplicate logic in pair_controller

ericworkman committed:
Move tests from controllers to pair_maker_test

ericworkman committed:
Add migration to handle adding existing user pairs to an existing group

ericworkman committed:
Remove accidental can? for Projects and :index action

ericworkman committed:
Correctly filter out non-user pairs in pair controller

ericworkman commented:
@mbramson Review please. 
Here's what the current /pairs looks like for a user:
<img width="747" alt="screen shot 2017-02-16 at 5 04 30 pm" src="https://cloud.githubusercontent.com/assets/4420576/23043428/36a51d7e-f46a-11e6-9b25-88cef6a73d47.png">
Versus the "The Oh Yeahs" group's pairs for the same period:
<img width="742" alt="screen shot 2017-02-16 at 5 04 46 pm" src="https://cloud.githubusercontent.com/assets/4420576/23043456/4d7273ee-f46a-11e6-88c4-3889f2ab4215.png">

The added migration will associate all current pairs to the only existing group in prod. For dev environments it could be a little weird if the first group isn't the right group for users. Worst case is that the dev db needs to be cleaned up. Any suggestions for improving that migration so that it is smarter?

Should we add the group's name to the group pair page?

mbramson commented on web/controllers/group_pair_controller.ex:
```diff
@@ -0,0 +1,44 @@
+  import Pairmotron.ControllerHelpers
+  alias Pairmotron.{PairMaker, Group}
+
+  plug :load_and_authorize_resource, model: Group, only: [:index, :show]
```
Why :index in addition to :show? There's no can? protocol defined and no index action on this controller.

ericworkman replied:
Oversight 🙈

mbramson commented on web/controllers/pair_controller.ex:
```diff
@@ -2,108 +2,36 @@ defmodule Pairmotron.PairController do
+               member_ids = p.users |> Enum.map(&(&1.id))
+               Enum.member?(member_ids, user.id)
+             end)
+      end)
```
Can we be explicit with our temp variables? So `group` instead of `g` and `pair` instead of `p`?

ericworkman replied:
Sure thing

mbramson commented on web/controllers/pair_controller.ex:
```diff
@@ -2,108 +2,36 @@ defmodule Pairmotron.PairController do
+          |> Enum.filter(fn(p) ->
+               member_ids = p.users |> Enum.map(&(&1.id))
+               Enum.member?(member_ids, user.id)
+             end)
```
Could we extract lines 29 through 33 to their own method called something like `users_pair_for_group(user, year, week, group)`. This code is about as elegant as it can be without extraction, but I think extraction would improve the ease of understanding both `fetch_pairs_per_group` and the current in-line block to be extracted.

ericworkman replied:
Sounds good to me

mbramson commented on lib/abilities.ex:
```diff
@@ -9,6 +9,12 @@ defimpl Canada.Can, for: Pairmotron.User do
+  def can?(user = %User{}, :show, group = %Group{}) do
+    group = group |> Pairmotron.Repo.preload(:users)
+    group.users
+      |> Enum.any?(fn guser -> guser.id == user.id end)
```
Could lines 14-15 be simplified to `Enum.member?(group.users, user)`

mbramson replied:
Also, I'm kind of worried about this being attached to the `%Group{}` can? when it is used on `GroupPairController`, as the `GroupController` uses the `%Group{}` can? implementation. I realize that they use different actions and so this fits nicely in here for now, but what if I decide that a Group :show action should be visible to someone different and want to expand the can? implementation?

Practical example of such an expansion includes maybe wanting users that have received invites for groups to be able to view the :show route in addition to only users in those groups (say if the group were private). You wouldn't want to also give invited users the option see the pairs for that group (maybe? potentially? just an example). In that case there would be collision for the two :show can? implementations, and I would think the one actually going to the Group itself's show action should win out.

Unfortunately, this is a limitation of Canary/Canada. I think that this permission check should be in the GroupPairController show action itself :(.

ericworkman replied:
> Could lines 14-15 be simplified to `Enum.member?(group.users, user)`

No unfortunately. That was the cause of the last bug I found. The user and the group.users have to match exactly, including preloaded associations and other things, for `Enum.member?` to work.

> I think that this permission check should be in the GroupPairController show action itself :(.

I leaned that way initially too. It fit perfectly as you say, but could be the cause of some trouble later. Want me to remove this now?

mbramson replied:
Ya I think it could also cause confusion down the line. Someone might not know where that rule is actually applied, and be misled about how authorization works for groups.

mbramson commented on web/router.ex:
```diff
@@ -43,20 +43,23 @@ defmodule Pairmotron.Router do
+
+    get "/groups/:id/pairs", GroupPairController, :show
+    get "/groups/:id/pairs/:year/:week", GroupPairController, :show
+    delete "/groups/:group_id/pairs/:year/:week", GroupPairController, :delete
```
The name of the group_id variable is different for these three routes. I think `:group_id` makes the most sense, but they should probably be consistent.

ericworkman replied:
oversight, thanks

mbramson commented on web/templates/group_pair/index.html.eex:
```diff
@@ -0,0 +1,20 @@
+  <div class="col-md-8 col-sm-12 col-xs-12">
+    <ul class="list-group">
+      <%= for pair <- @pairs do %>
+        <%= render Pairmotron.SharedView, "pair.html", pair: pair, conn: @conn %>
```
Actually did not realize you could plug and play views like that. Thought it was templates only. Awesome!

mbramson commented on test/controllers/group_pair_controller_test.exs:
```diff
@@ -0,0 +1,101 @@
+      conn = get(conn, "/groups/#{group.id}/pairs")
+      assert html_response(conn, 200) =~ user.name
+    end
+
```
What about if there are no users in the group at all? Weirdly enough, this is a possible state right now. You can be the owner of the group and not actually in the users association.

ericworkman replied:
Then you won't even see these pages. Want me to add a test case around this?

mbramson replied:
Up to you. It's definitely an edge case.

ericworkman replied:
I feel like this is handled by the tests of non-group membership access and inactive users.

mbramson commented on test/controllers/group_pair_controller_test.exs:
```diff
@@ -0,0 +1,101 @@
+    test "cannot access group pairs for a specific period", %{conn: conn, group: group} do
+      conn = get conn, group_pair_path(conn, :show, group.id, 2000, 1)
+      assert redirected_to(conn) == pair_path(conn, :show, 2000, 1)
+    end
```
This might not warrant a test, but should an admin be able to repairify a group that they're not in? I'm not sure what the answer is.

ericworkman replied:
Good call. I'll add a test case to the `describe "while authenticated as admin" do` section.

ericworkman replied:
Now handled in the "while authenticated as admin" section. 👍

mbramson reviewed (changes requested):
Awesome work! I have a few small things that I think should be addressed, but otherwise this looked great!

mbramson commented:
And also, that's a good point. I think that we should add the group name to the group specific pairs page.

ericworkman committed:
Improve admin/nonadmin repairifying controller tests

ericworkman committed:
Refactor pair controller's pair fetching function for readability

ericworkman committed:
Refactor for duplicate logic and non-can authorization

ericworkman committed:
Include group's name on group pairs page

ericworkman committed:
Add missing divs to pair index so it matches group_pair

ericworkman committed:
Make :delete's parameters match the others' parameters

ericworkman commented:
@mbramson Added the group name to the group_pair page. Ready for more review.

mbramson commented on web/router.ex:
```diff
@@ -43,20 +43,23 @@ defmodule Pairmotron.Router do
+
+    get "/groups/:id/pairs", GroupPairController, :show
+    get "/groups/:id/pairs/:year/:week", GroupPairController, :show
+    delete "/groups/:id/pairs/:year/:week", GroupPairController, :delete
```
Why `:id` instead of `:group_id`? As these are routes for `Pair`s, I'd think `:id` would be reserved for a specific pair (if there was ever some route for updating a specific pair for some reason). Therefore, I'd tihnk that `:group_id` would be more appropriate here for the route.

ericworkman replied:
The `:id` is what allowed canary to match on a `%Group{}`. I agree that `:group_id` is more specific, but I think `:id` matches the general pattern. These are really routes for groups first, then routes for pairs. If we could go the embedded resources way, it would still come through as `:id`.

mbramson reviewed (approved):
One thought on the name of a variable in some routes, but otherwise LGTM! Fantastic work!
